Sources/NIOFileSystem/FileChunks.swift

```diff
@@ -260,87 +334,216 @@ private struct ProducerState: Sendable {
+            switch producingState.activityState {
+            case .inactive:
+                producingState.activityState = .active
+                self.state = .producing(producingState)
+                return .performRead
+            case .active:
+                producingState.activityState = .activeAndQueued
+                self.state = .pausedProducing(producingState)
+                return .stop
+            case .activeAndQueued:
+                return .stop
+            }
+
+        case .done:
+            return .stop
+        }
+    }
+
+    /// Actions which may be taken after a more data is produced.
+    ///
+    /// Either go on to read more or stop.
+    internal enum PerformedProduceMoreAction {
+        case readMore
+        case stop
+    }
+
+    /// Update state to reflect that a more data has been produced.
+    mutating func performedProduceMore() -> PerformedProduceMoreAction {
+        switch self.state {
+        case .producing(var producingState):
+            let oldActivityState = producingState.activityState
+
+            producingState.activityState = .inactive
+            self.state = .producing(producingState)
+
+            switch oldActivityState {
+            case .inactive:
+                preconditionFailure()
+            case .active, .activeAndQueued:
+                return .readMore
+            }
+
+        case .pausedProducing(var producingState):
+            let oldActivityState = producingState.activityState
+
+            producingState.activityState = .inactive
+            self.state = .pausedProducing(producingState)
+
+            switch oldActivityState {
+            case .inactive:
+                preconditionFailure()
+            case .active:
+                return .stop
+            case .activeAndQueued:
+                return .readMore
+            }
+
+        case .done:
+            return .stop
+        }
+    }
+
+    internal enum RangeUpdateOutcome {
+        case moreToRead
+        case cannotReadMore
+    }
+
+    mutating func activeThreadPool() -> NIOThreadPool? {
         switch self.state {
-        case let .producing(state):
-            return state.handle.threadPool
+        case .producing(let producingState), .pausedProducing(let producingState):
+            return producingState.handle.threadPool
         case .done:
             return nil
         }
     }
 
-    mutating func produceMore() -> Result<(FileDescriptor, Range<Int64>?)?, FileSystemError> {
+    mutating func fileReadingState() -> Result<(FileDescriptor, Range<Int64>?)?, FileSystemError> {
         switch self.state {
-        case let .producing(state):
-            if let descriptor = state.handle.descriptorIfAvailable() {
-                return .success((descriptor, state.range))
-            } else {
+        case .producing(let producingState), .pausedProducing(let producingState):
+            guard let descriptor = producingState.handle.descriptorIfAvailable() else {
```

Comment:
nit: I don't think a `guard` is any clearer here as the exit on the failure path isn't earlier than on the success path. IMO if both paths immediately return then `if`/`else` is clearer.